pkg/shim/util_unix.go

```diff
@@ -128,8 +129,12 @@ func NewSocket(address string) (*net.UnixListener, error) {
 	}
 
 	if !isAbstract {
+		// Ensure the state directory ("/run/containerd/") gets created with the correct permissions
+		if err := os.MkdirAll(socketRoot, stateDirPerm); err != nil {
+			return nil, err
+		}
 		if err := os.MkdirAll(filepath.Dir(path), perm); err != nil {
```

Comment:
So.. hm.. wouldn't `filepath.Dir(path)` be the same as the code above? Is the issue here that `perm` is used, which should be for the _socket_ permissions, not for the _directory_ in which it lives?

Reply:
And I'm even starting to wonder if this `MkdirAll` is redundant altogether, as  it looks like the `containerd` daemon holds the responsibility of setting up all the required directories as early as possible; It looks like `CreateTopLevelDirectories` is responsible for creating these; https://github.com/containerd/containerd/blob/ea3e1d9613f257290cffcd5522ad98afe1ed336d/cmd/containerd/server/server.go#L74-L90 Which is executed as part of `containerd`'s main process (and a comment "Make sure top-level directories are created early."); https://github.com/containerd/containerd/blob/ea3e1d9613f257290cffcd5522ad98afe1ed336d/cmd/containerd/command/main.go#L157-L158

Reply:
> So.. hm.. wouldn't filepath.Dir(path) be the same as the code above? Is the issue here that perm is used, which should be for the socket permissions, not for the directory in which it lives?

Path includes the `s` as the intermediary directory : https://github.com/containerd/containerd/blob/main/pkg/shim/util_unix.go#L89

Reply:
Oh! I saw that, and then forgot 🙈 thanks!